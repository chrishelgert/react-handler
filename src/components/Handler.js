import React, { PropTypes } from 'react';
import { LoadingMessage, ErrorMessage, EmptyMessage } from './Messages';
import isEmpty from '../utils/isEmpty';

/**
 * The Handler handles loading, error and empty state
 * or display the given Component.
 */
const Handler = ({
  LoadingComponent,
  ErrorComponent,
  EmptyComponent,
  Component,
  errorMsg,
  emptyMsg,
  emptyPropertyKey,
  componentProps = {},
  loading = true,
}) => {
  if (loading) {
    return LoadingComponent ? <LoadingComponent /> : <LoadingMessage />;
  }

  if (errorMsg) {
    if (ErrorComponent) {
      return <ErrorComponent>{errorMsg}</ErrorComponent>;
    }

    return <ErrorMessage>{errorMsg}</ErrorMessage>;
  }

  if (isEmpty(emptyPropertyKey, componentProps)) {
    if (EmptyComponent) {
      return <EmptyComponent>{emptyMsg}</EmptyComponent>;
    }

    return <EmptyMessage>{emptyMsg}</EmptyMessage>;
  }

  return <Component {...componentProps} />;
};

Handler.propTypes = {
  LoadingComponent: PropTypes.func,
  ErrorComponent: PropTypes.func,
  EmptyComponent: PropTypes.func,
  Component: PropTypes.func.isRequired,
  errorMsg: PropTypes.string,
  emptyMsg: PropTypes.string,
  emptyPropertyKey: PropTypes.string,
  componentProps: PropTypes.object,
  loading: PropTypes.bool,
};

export default Handler;
