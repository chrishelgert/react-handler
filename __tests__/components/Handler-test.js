import React, { PropTypes } from 'react';
import renderer from 'react-test-renderer';
import Handler from '../../src/components/Handler';

const LoadingComponent = () => <h2 className="loading">loading...</h2>;
const ErrorComponent = ({ children }) => <h2 className="error">{children}</h2>;
ErrorComponent.propTypes = { children: PropTypes.string };
const EmptyComponent = ({ children }) => <h2 className="empty">{children}</h2>;
EmptyComponent.propTypes = { children: PropTypes.string };
const Component = () => <div>Test</div>;
const errorMsg = 'error..';
const emptyMsg = 'empty..';

describe('Handler', () => {
  describe('loading-state', () => {
    it('should return the LoadingComponent, when it is set', () => {
      const tree = renderer.create(
        <Handler
          LoadingComponent={LoadingComponent}
          Component={Component}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should return the LoadingMessage', () => {
      const tree = renderer.create(
        <Handler Component={Component} />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('error-state', () => {
    it('should return the ErrorComponent, when it is set', () => {
      const tree = renderer.create(
        <Handler
          loading={false}
          errorMsg={errorMsg}
          ErrorComponent={ErrorComponent}
          Component={Component}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should return the ErrorMessage', () => {
      const tree = renderer.create(
        <Handler
          loading={false}
          errorMsg={errorMsg}
          Component={Component}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('empty-state', () => {
    it('should return the EmptyComponent, when it is set', () => {
      const tree = renderer.create(
        <Handler
          loading={false}
          EmptyComponent={EmptyComponent}
          emptyPropertyKey="list"
          emptyMsg={emptyMsg}
          componentProps={{ list: [] }}
          Component={Component}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should return the EmptyMessage', () => {
      const tree = renderer.create(
        <Handler
          loading={false}
          emptyPropertyKey="list"
          emptyMsg={emptyMsg}
          componentProps={{ list: [] }}
          Component={Component}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe('Component', () => {
    it('should return the Component, when there is no error, loading false and no emptyPropertyKey', () => {
      const tree = renderer.create(
        <Handler
          LoadingComponent={LoadingComponent}
          Component={Component}
          loading={false}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should return the Component, when there is no error, loading false and the entry filled', () => {
      const tree = renderer.create(
        <Handler
          LoadingComponent={LoadingComponent}
          Component={Component}
          componentProps={{ list: ['abc'] }}
          emptyPropertyKey="list"
          loading={false}
        />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
