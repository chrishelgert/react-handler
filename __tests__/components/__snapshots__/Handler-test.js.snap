exports[`Handler Component should return the Component, when there is no error, loading false and no emptyPropertyKey 1`] = `
<div>
  Test
</div>
`;

exports[`Handler Component should return the Component, when there is no error, loading false and the entry filled 1`] = `
<div>
  Test
</div>
`;

exports[`Handler empty-state should return the EmptyComponent, when it is set 1`] = `
<h2
  className="empty">
  empty..
</h2>
`;

exports[`Handler empty-state should return the EmptyMessage 1`] = `
<div
  className="handler--empty">
  empty..
</div>
`;

exports[`Handler error-state should return the ErrorComponent, when it is set 1`] = `
<h2
  className="error">
  error..
</h2>
`;

exports[`Handler error-state should return the ErrorMessage 1`] = `
<div
  className="handler--error">
  error..
</div>
`;

exports[`Handler loading-state should return the LoadingComponent, when it is set 1`] = `
<h2
  className="loading">
  loading...
</h2>
`;

exports[`Handler loading-state should return the LoadingMessage 1`] = `
<div
  className="handler--loading">
  loading...
</div>
`;
